text/0000-cargo-workspace.md

```diff
@@ -0,0 +1,345 @@
+  root and encompass the dependencies (and dev-dependencies) for all packages
+  in the workspace.
+
+With workspaces, Cargo can now solve the problems set forth in the motivation
+section. Next, however, workspaces need to be defined. In the spirit of much of
+the rest of Cargo's configuration today this will largely be automatic for
+conventional project layouts but will have explicit controls for configuration.
+
+### New manifest keys
+
+First, let's look at the new manifest keys which will be added to `Cargo.toml`:
+
+```toml
+[workspace]
+members = ["relative/path/to/child1", "../child2"]
+
+# or ...
+
+[package]
+workspace = "../foo"
+```
+
+The root of a workspace, indicated by the presence of `[workspace]`, is
+responsible for defining the entire workspace (listing all members).
+This example here means that two extra crates will members of the workspace
+(which also includes the root).
+
+The `package.workspace` key is used to point at a workspace root. For
+example this Cargo.toml indicates that the Cargo.toml in `../foo` is the
+workspace root that this package is a member of.
+
+These keys are mutually exclusive when applied in `Cargo.toml`. A crate may
+*either* specify `package.workspace` or specify `[workspace]`. That is, a
+crate cannot both be a root in a workspace (contain `[workspace]`) and also be
+member of another workspace (contain `package.workspace`).
+
+### "Virtual" `Cargo.toml`
+
+A good number of projects do not necessarily have a "root `Cargo.toml`" which is
+an appropriate root for a workspace. To accommodate these projects and allow for
+the output of a workspace to be configured regardless of where crates are
+located, Cargo will now allow for "virtual manifest" files. These manifests will
+currently **only** contains the `[workspace]` table and will notably be lacking
+a `[project]` or `[package]` top level key.
+
+Cargo will for the time being disallow many commands against a virtual manifest,
+for example `cargo build` will be rejected. Arguments that take a package,
+however, such as `cargo test -p foo` will be allowed. Workspaces can eventually
+get extended with `--all` flags so in a workspace root you could execute
+`cargo build --all` to compile all crates.
+
+### Validating a workspace
+
+A workspace is valid if these two properties hold:
+
+1. A workspace has only one root crate (that with `[workspace]` in
+   `Cargo.toml`).
+2. All workspace crates defined in `workspace.members` point back to the
+   workspace root with `package.workspace`.
+
+While the restriction of one-root-per workspace may make sense, the restriction
+of crates pointing back to the root may not. If, however, this restriction were
+not in place then the set of crates in a workspace may differ depending on
+which crate it was viewed from. For example if workspace root A includes B then
+it will think B is in A's workspace. If, however, B does ont point back to A,
```

Comment:
Typo: `ont` should be `not`.

(Please tell me if I'm not allowed to point out typos here)